test/geopackage/WKB_Test.cpp

```diff
@@ -0,0 +1,262 @@
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x40
+        };
+    }
+
+    void TearDown() override {}
+    
+    std::map<std::string, std::vector<uint8_t>> wkb_bytes;
+};
+
+TEST_F(WKB_Test, wkb_point_test) // also tests endianness
+{  
+    const auto geom_big    = boost::get<wkb::point_t>(wkb::read(this->wkb_bytes["point_big"]));
+    const auto geom_little = boost::get<wkb::point_t>(wkb::read(this->wkb_bytes["point"]));
+    EXPECT_NEAR(geom_big.get<0>(), geom_little.get<0>(), 0.000001);
+    EXPECT_NEAR(geom_big.get<1>(), geom_little.get<1>(), 0.000001);
+    EXPECT_NEAR(geom_big.get<0>(), 2.0, 0.000001);
+    EXPECT_NEAR(geom_big.get<1>(), 4.0, 0.000001);
+    EXPECT_NEAR(geom_big.get<0>(), 2.0, 0.000001);
+    EXPECT_NEAR(geom_big.get<1>(), 4.0, 0.000001);
+}
+
+TEST_F(WKB_Test, wkb_linestring_test)
+{
+    const wkb::geometry geom = wkb::read(this->wkb_bytes["linestring"]);
+    EXPECT_EQ(geom.which() + 1, 2);
+
+    const wkb::linestring_t& line = boost::get<wkb::linestring_t>(geom);
+    const std::vector<std::pair<double, double>> expected_coordinates = {
+        {30, 10}, {10, 30}, {40, 40}
+    };
+
+    for (int i = 0; i < expected_coordinates.size(); i++) {
+        EXPECT_NEAR(line[i].get<0>(), expected_coordinates[i].first, 0.0001);
+        EXPECT_NEAR(line[i].get<1>(), expected_coordinates[i].second, 0.0001);
+    }
+}
+
+TEST_F(WKB_Test, wkb_polygon_test)
+{    
+    const wkb::geometry geom = wkb::read(this->wkb_bytes["polygon"]);
+    EXPECT_EQ(geom.which() + 1, 3); // +1 since variant.which() is 0-based
+
+    const wkb::polygon_t& poly = boost::get<wkb::polygon_t>(geom);
+    const std::vector<std::pair<double, double>> expected_coordinates = {
+        {10.689, -25.092},
+        {34.595, -20.170},
+        {38.814, -35.639},
+        {13.502, -39.155},
+        {10.689, -25.092}
+    };
+
+    for (int i = 0; i < expected_coordinates.size(); i++) {
+        EXPECT_NEAR(poly.outer().at(i).get<0>(), expected_coordinates[i].first, 0.0001);
+        EXPECT_NEAR(poly.outer().at(i).get<1>(), expected_coordinates[i].second, 0.0001);
+    }
+}
+
+TEST_F(WKB_Test, wkb_polygon_with_holes_test)
+{
+    const wkb::geometry geom = wkb::read(this->wkb_bytes["polygon_with_holes"]);
+    EXPECT_EQ(geom.which() + 1, 3);
+
+    const wkb::polygon_t& poly = boost::get<wkb::polygon_t>(geom);
+    ASSERT_EQ(poly.inners().size(), 1);
+
+    const std::vector<std::pair<double, double>> expected_outer = {
+        {0, 0},
+        {5, 0},
+        {5, 5},
+        {0, 5},
+        {0, 0}
+    };
+
+    for (int i = 0; i < expected_outer.size(); i++) {
+        EXPECT_NEAR(poly.outer()[i].get<0>(), expected_outer[i].first, 0.0001);
+        EXPECT_NEAR(poly.outer()[i].get<1>(), expected_outer[i].second, 0.0001);
+    }
+
+    const std::vector<std::pair<double, double>> expected_inner = {
+        {2, 2},
+        {3, 2},
+        {3, 3},
+        {2, 3},
+        {2, 2}
+    };
```

Comment:
I'm not sure if I fully understood what I read, but aren't inner polygon boundaries supposed to run clockwise, to maintain orientation?

Reply:
You're correct. That is the case typically (or at least all polygons in a given implementation should orient the same way). Boost doesn't care for orientation (as far as I know), so order is given as read. Most geometry writers will handle orientation and (sometimes) topology validation, so that felt out of scope to handle in our case.

In this test case specifically, we aren't checking for topological correctness, just value correctness.

Reply:
Is there anything in the broader application that relies on that orientation? If so, is that being checked at a higher level, or does it expect the WKB (and geojson) readers to ensure the validity of what they return?

Reply:
If by broader application you mean `ngen`, then no I don't think so. The geometry validity is a hydrofabric responsibility I think, so that isn't something we would need to be concerned with. As far as the validity of outputs from WKB/geopackage/geojson writers, assuming our inputs are the hydrofabric, then we are relying on the validity of GDAL, which I think is good enough to not worry about.

@hellkite500 might have a better answer than me (because I could be completely wrong haha).

Reply:
This is an interesting question, especially when it comes to topology validation.  I can definitely see a use case that takes a catchment polygon (or set of polygons) and wants to do something like create a dynamic mesh using the boundaries as pilot points.  In this case, topology is quite important.  Nothing in our current and short/medium term goals, however, is going to be doing something that sophisticated.  

Assuming the inputs are always generated by a specific workflow/tool isn't a great assumption either.  The general intent is that anyone can generate/produce a hydrofabric of their liking and we can simulate on it.  I would say handling this validation is likely out of scope for this PR, but it shouldn't be entirely ignored.  I think some in code comments stating the assumptions are worthwhile, and probably an issue to track/document our ultimate decision on how much validation we perform is warranted.

Reply:
I think creating an issue on this topic is worthwhile.

My assumption is less on whether its generated by a specific workflow/tool, but more of if the inputs are generated using _standard-conforming_ writers and representations.

If the inputted geometry is already invalid, that shouldn't be an `ngen` responsibility, since that requires modifying the geometry before processing, which also requires inferring the user's intent (i.e. we fix a divide's topology, but now its covering is inaccurate -- in this case a warning might be useful).

However, if _we_ are generating geometry, then I think that is a different conversation. At that point, validation should be considered (if not necessary).

Regardless, I think it's something worth discussing though! 😄

Reply:
Given the breadth of possible sources of data, I'm strongly inclined toward saying that we should implement internal validation of whatever we're reading, and enforce it sooner rather than later. Otherwise, we're going to end up committed to supporting those invalid inputs, even as we want to do things that would benefit from stricter assumptions, and so we'll have to write clean-up code.

None of that argues for doing it in this PR, though. Please open an issue for it and link it here.

Reply:
Also, please make these test cases strictly valid inputs

Reply:
updated in bd30664.

Reply:
Trust but validate....I think they call that `0 trust` these days haha